### PR TITLE
Tests for TAC Participation Skill

### PR DESCRIPTION
**Pull request by 5A11**
This PR adds tests for tac participation skill.

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/test_packages/test_skills/test_tac_participation/test_behaviours.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+        # _search_for_tac
+        message = self.get_message_from_outbox()
+        has_attributes, error_str = self.message_has_attributes(
+            actual_message=message,
+            message_type=OefSearchMessage,
+            performative=OefSearchMessage.Performative.SEARCH_SERVICES,
+            to=self.skill.skill_context.search_service_address,
+            sender=self.skill.skill_context.agent_address
+            + "_"
+            + str(self.skill.skill_context.skill_id),
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_participation/test_behaviours.py:100]: we need to fix this

File: tests/test_packages/test_skills/test_tac_participation/test_behaviours.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+            )
+            message = self.get_message_from_outbox()
+            has_attributes, error_str = self.message_has_attributes(
+                actual_message=message,
+                message_type=TacMessage,
+                performative=TacMessage.Performative.TRANSACTION,
+                to=COUNTERPARTY_ADDRESS,
+                sender=self.skill.skill_context.agent_address,
+                target=tac_dialogue.last_message.message_id
+                - 2
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_participation/test_behaviours.py:244]: define constant at start of test class to make it easier to read

File: tests/test_packages/test_skills/test_tac_participation/test_behaviours.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                counterparty_address=self.terms[count].counterparty_address,
+                amount_by_currency_id=self.terms[count].amount_by_currency_id,
+                fee_by_currency_id=self.terms[count].fee_by_currency_id,
+                quantities_by_good_id=self.terms[count].quantities_by_good_id,
+                sender_signature=self.sender_signatures[count],
+                counterparty_signature=self.counterparty_signatures[count],
+                nonce=self.terms[count].nonce,
+            )
+            assert has_attributes, error_str
+            count += 1
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_participation/test_behaviours.py:258]: what's that for? tests run in any order, so don't assume order across methods

File: tests/test_packages/test_skills/test_tac_participation/test_dialogues.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            counterparty=COUNTERPARTY_ADDRESS,
+            performative=OefSearchMessage.Performative.SEARCH_SERVICES,
+            query="some_query",
+        )
+        assert dialogue.role == OefSearchDialogue.Role.AGENT
+        assert (
+            dialogue.self_address
+            == self.skill.skill_context.agent_address
+            + "_"
+            + str(self.skill.skill_context.skill_id)
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_participation/test_dialogues.py:70]: we need to fix this

File: tests/test_packages/test_skills/test_tac_participation/test_dialogues.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            performative=StateUpdateMessage.Performative.INITIALIZE,
+            exchange_params_by_currency_id={"some_currency_id": 1.0},
+            utility_params_by_good_id={"some_good_id": 2.0},
+            amount_by_currency_id={"some_currency_id": 10},
+            quantities_by_good_id={"some_good_id": 5},
+        )
+        assert dialogue.role == StateUpdateDialogue.Role.SKILL
+        assert (
+            dialogue.self_address
+            == self.skill.skill_context.agent_address
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_participation/test_dialogues.py:86]: see above

File: tests/test_packages/test_skills/test_tac_participation/test_handlers.py [hunk 6]
```diff
@@ -0,0 +1,917 @@
...
+            dialogues=self.oef_dialogues, messages=self.list_of_messages[:1],
+        )
+        incoming_message = cast(
+            OefSearchMessage,
+            self.build_incoming_message_for_skill_dialogue(
+                dialogue=oef_dialogue,
+                performative=OefSearchMessage.Performative.OEF_ERROR,
+                to=self.skill.skill_context.agent_address
+                + "_"
+                + str(self.skill.skill_context.skill_id),
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_participation/test_handlers.py:113]: see above

File: tests/test_packages/test_skills/test_tac_participation/test_handlers.py [hunk 7]
```diff
@@ -0,0 +1,917 @@
...
+            dialogues=self.oef_dialogues, messages=self.list_of_messages[:1],
+        )
+        incoming_message = cast(
+            OefSearchMessage,
+            self.build_incoming_message_for_skill_dialogue(
+                dialogue=oef_dialogue,
+                performative=OefSearchMessage.Performative.SEARCH_RESULT,
+                to=self.skill.skill_context.agent_address
+                + "_"
+                + str(self.skill.skill_context.skill_id),
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_participation/test_handlers.py:141]: see above

File: tests/test_packages/test_skills/test_tac_participation/test_handlers.py [hunk 8]
```diff
@@ -0,0 +1,917 @@
...
+
+        self.assert_quantity_in_decision_making_queue(1)
+        has_attributes, error_str = self.message_has_attributes(
+            actual_message=self.get_message_from_decision_maker_inbox(),
+            message_type=StateUpdateMessage,
+            performative=StateUpdateMessage.Performative.INITIALIZE,
+            to=self.skill.skill_context.decision_maker_address,
+            sender=self.skill.skill_context.agent_address
+            + "_"
+            + str(self.skill.skill_context.skill_id),
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_participation/test_handlers.py:519]: see above

File: tests/test_packages/test_skills/test_tac_participation/test_handlers.py [hunk 9]
```diff
@@ -0,0 +1,917 @@
...
+
+        self.assert_quantity_in_decision_making_queue(1)
+        has_attributes, error_str = self.message_has_attributes(
+            actual_message=self.get_message_from_decision_maker_inbox(),
+            message_type=StateUpdateMessage,
+            performative=StateUpdateMessage.Performative.APPLY,
+            to=self.skill.skill_context.decision_maker_address,
+            sender=self.skill.skill_context.agent_address
+            + "_"
+            + str(self.skill.skill_context.skill_id),
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/test_tac_participation/test_handlers.py:805]: see above

File: tox.ini [hunk 10]
```diff
@@ -36,7 +36,7 @@ deps =
...
+commands = pytest -rfE --doctest-modules aea packages/fetchai/connections packages/fetchai/protocols packages/fetchai/skills/generic_buyer packages/fetchai/skills/generic_seller packages/fetchai/skills/tac_control packages/fetchai/skills/tac_control_contract packages/fetchai/skills/tac_participation tests/ --cov-report=html --cov-report=xml --cov-report=term --cov-report=term-missing --cov=aea --cov=packages/fetchai/connections --cov=packages/fetchai/protocols --cov=packages/fetchai/skills/generic_buyer --cov=packages/fetchai/skills/generic_seller --cov=packages/fetchai/skills/tac_control --cov=packages/fetchai/skills/tac_control_contract --cov=packages/fetchai/skills/tac_participation --cov-config=.coveragerc {posargs}
```
Comment (DavidMinarsch) [inline tox.ini:39]: Also add in Makefile commands

File: tests/test_packages/test_skills/test_tac_participation/test_behaviours.py (hunk 2 above)
Comment (5A11) [reply tests/test_packages/test_skills/test_tac_participation/test_behaviours.py:244]: Done

File: tests/test_packages/test_skills/test_tac_participation/test_behaviours.py (hunk 3 above)
Comment (5A11) [reply tests/test_packages/test_skills/test_tac_participation/test_behaviours.py:258]: Yes. That is for the while loop. 

The skill code sends a message for every tx in shared_state. So here, the while loops through transactions and validates every message in the outbox.

File: tox.ini (hunk 10 above)
Comment (5A11) [reply tox.ini:39]: Done (double check the format is correct please)

File: tests/test_packages/test_skills/test_tac_participation/test_behaviours.py (hunk 1 above)
Comment (5A11) [reply tests/test_packages/test_skills/test_tac_participation/test_behaviours.py:100]: How? Do you mean changing it to the usual `self.skill.skill_context.agent_address`, without the 
```
+ "_"
+ str(self.skill.skill_context.skill_id)
```
part?

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**